stl/inc/expected

```diff
@@ -685,6 +689,303 @@ public:
+            "N4928 [expected.object.monadic]/7.");
+
+        if (_Has_value) {
+            return _STD invoke(_STD forward<_Fn>(_Func), _STD move(_Value));
+        } else {
+            return _Uty(unexpect, _STD move(_Unexpected));
+        }
+    }
+
+    template <class _Fn>
+        requires is_move_constructible_v<_Err>
+    constexpr auto and_then(_Fn&& _Func) const&& {
+        using _Uty = remove_cvref_t<invoke_result_t<_Fn, const _Ty&&>>;
+
+        static_assert(_Is_specialization_v<_Uty, expected>,
+            "expected<T,E>::and_then(F) requires the return type of F to be a specialization of expected "
+            "N4928 [expected.object.monadic]/7.");
+
+        if (_Has_value) {
+            return _STD invoke(_STD forward<_Fn>(_Func), _STD move(_Value));
+        } else {
+            return _Uty(unexpect, _STD move(_Unexpected));
+        }
+    }
+
+    template <class _Fn>
+        requires is_copy_constructible_v<_Ty>
+    constexpr auto or_else(_Fn&& _Func) & {
+        using _Uty = remove_cvref_t<invoke_result_t<_Fn, _Err&>>;
+
+        static_assert(_Is_specialization_v<_Uty, expected>,
+            "expected<T,E>::or_else(F) requires the return type of F to be a specialization of expected "
+            "N4928 [expected.object.monadic]/11.");
+        static_assert(is_same_v<typename _Uty::value_type, _Ty>,
+            "expected<T,E>::or_else(F) requires the value type of the return type of F to be T"
+            "N4928 [expected.object.monadic]/11.");
+
+        if (_Has_value) {
+            return _Uty(in_place, _Value);
+        } else {
+            return _STD invoke(_STD forward<_Fn>(_Func), _Unexpected);
+        }
+    }
+
+    template <class _Fn>
+        requires is_copy_constructible_v<_Ty>
+    constexpr auto or_else(_Fn&& _Func) const& {
+        using _Uty = remove_cvref_t<invoke_result_t<_Fn, const _Err&>>;
+
+        static_assert(_Is_specialization_v<_Uty, expected>,
+            "expected<T,E>::or_else(F) requires the return type of F to be a specialization of expected "
+            "N4928 [expected.object.monadic]/11.");
+        static_assert(is_same_v<typename _Uty::value_type, _Ty>,
+            "expected<T,E>::or_else(F) requires the value type of the return type of F to be T"
+            "N4928 [expected.object.monadic]/11.");
+
+        if (_Has_value) {
+            return _Uty(in_place, _Value);
+        } else {
+            return _STD invoke(_STD forward<_Fn>(_Func), _Unexpected);
+        }
+    }
+
+    template <class _Fn>
+        requires is_move_constructible_v<_Ty>
+    constexpr auto or_else(_Fn&& _Func) && {
+        using _Uty = remove_cvref_t<invoke_result_t<_Fn, _Err&&>>;
+
+        static_assert(_Is_specialization_v<_Uty, expected>,
+            "expected<T,E>::or_else(F) requires the return type of F to be a specialization of expected "
+            "N4928 [expected.object.monadic]/11.");
+        static_assert(is_same_v<typename _Uty::value_type, _Ty>,

```

Comment:
This `typename` shouldn't be required anymore.
https://godbolt.org/z/TnvrdPKE7

Reply:
Clang doesn't implement p0634r3 in a released version yet (it'll be in v16) so I thought it best to keep it. Your Godbolt link only compiles because `Foo<int>::value_type` isn't a dependent name. If you make it dependent, Clang fails to compile it: https://godbolt.org/z/brGs9efxY

Reply:
Thanks for your reply! and +1